pkg/component/controller/etcd_member_reconciler.go

```diff
@@ -0,0 +1,352 @@
+			if err != nil {
+				log.WithError(err).Error("failed to update member status")
+			}
+			return nil
+		} else {
+			return err
+		}
+	}
+
+	em.PeerAddress = e.etcdConfig.PeerAddress
+	em.MemberID = memberIDStr
+	em.Leave = false
+
+	log.Debug("EtcdMember object already exists, updating it")
+	// Update the object if it already exists
+	em, err = e.etcdMemberClient.Update(ctx, em, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	em.Status = etcdv1beta1.Status{
+		Conditions: []etcdv1beta1.JoinCondition{
+			{
+				Type:   etcdv1beta1.ConditionTypeJoined,
+				Status: etcdv1beta1.ConditionTrue,
+			},
+		},
+		ReconcileStatus: "",
+		Message:         "",
+	}
+	_, err = e.etcdMemberClient.UpdateStatus(ctx, em, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *EtcdMemberReconciler) reconcileMember(ctx context.Context, member *etcdv1beta1.EtcdMember) {
+	log := logrus.WithFields(logrus.Fields{
+		"component":   "etcdMemberReconciler",
+		"phase":       "reconcile",
+		"name":        member.Name,
+		"memberID":    member.MemberID,
+		"peerAddress": member.PeerAddress,
+	})
+
+	if !e.leaderElector.IsLeader() {
+		log.Debug("not the leader, skipping reconcile")
+		return
+	}
+
+	log.Debugf("reconciling EtcdMember: %+v", member)
+
+	if !member.Leave {
+		log.Debug("member not marked for leave, no action needed")
+		return
+	}
+
+	etcdClient, err := etcd.NewClient(e.k0sVars.CertRootDir, e.k0sVars.EtcdCertDir, e.etcdConfig)
+	if err != nil {
+		log.WithError(err).Warn("failed to create etcd client")
+		member.Status.ReconcileStatus = "Failed"
+		member.Status.Message = err.Error()
+		if _, err = e.etcdMemberClient.UpdateStatus(ctx, member, metav1.UpdateOptions{}); err != nil {
+			log.WithError(err).Error("failed to update member state")
+		}
+
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	// Verify that the member is actualy still present in etcd
+	members, err := etcdClient.ListMembers(ctx)
+	if err != nil {
+		member.Status.ReconcileStatus = "Failed"
+		member.Status.Message = err.Error()
+		if _, err = e.etcdMemberClient.UpdateStatus(ctx, member, metav1.UpdateOptions{}); err != nil {
+			log.WithError(err).Error("failed to update member state")
+		}
+
+		return
+	}
+
+	// Member marked for leave but no member found in etcd, mark for leaved
+	_, ok := members[member.Name]
+	if !ok {
+		log.Debug("member marked for leave but not in actual member list, updating state to reflect that")
+		member.Status.SetCondition(etcdv1beta1.ConditionTypeJoined, etcdv1beta1.ConditionFalse, member.Status.Message, time.Now())
+		member, err = e.etcdMemberClient.UpdateStatus(ctx, member, metav1.UpdateOptions{})
+		if err != nil {
+			log.WithError(err).Error("failed to update EtcdMember status")
+		}
+	}
+
+	if member.Status.GetCondition(etcdv1beta1.ConditionTypeJoined).Status == etcdv1beta1.ConditionFalse && !ok {
```

Comment:
Observed a panic during testing:

```text
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x2aaa037]

goroutine 432 [running]:
github.com/k0sproject/k0s/pkg/component/controller.(*EtcdMemberReconciler).reconcileMember(0xc000d89940, {0x40de860, 0xc000e077d0}, 0xc001c08f00)
        github.com/k0sproject/k0s/pkg/component/controller/etcd_member_reconciler.go:318 +0xe37
github.com/k0sproject/k0s/pkg/component/controller.(*EtcdMemberReconciler).Start.func1.2(0x0?)
        github.com/k0sproject/k0s/pkg/component/controller/etcd_member_reconciler.go:110 +0x25
github.com/k0sproject/k0s/pkg/kubernetes/watch.(*Watcher[...]).processWatchEvent(0xc0024976ac?, 0xc001d1d6f0?, 0x0)
        github.com/k0sproject/k0s/pkg/kubernetes/watch/watcher.go:340 +0x177
github.com/k0sproject/k0s/pkg/kubernetes/watch.(*Watcher[...]).watch(0x4116d20, {0x40de898, 0xc0006b5b30}, {0xc0001ddba8, 0x5}, 0xc001d1dcd0)
        github.com/k0sproject/k0s/pkg/kubernetes/watch/watcher.go:319 +0x3f1
github.com/k0sproject/k0s/pkg/kubernetes/watch.(*Watcher[...]).run(0x4116d20, {0x40de898, 0xc0006b5b30}, 0xc001d1dcd0)
        github.com/k0sproject/k0s/pkg/kubernetes/watch/watcher.go:247 +0x76
github.com/k0sproject/k0s/pkg/component/controller.(*EtcdMemberReconciler).Start.func1.(*Watcher[...]).Until.3()
        github.com/k0sproject/k0s/pkg/kubernetes/watch/watcher.go:198 +0x79
github.com/k0sproject/k0s/pkg/kubernetes/watch.retry({0x40de860, 0xc000e077d0}, 0xc0016601a0, 0xc001d1df80)
        github.com/k0sproject/k0s/pkg/kubernetes/watch/watcher.go:390 +0x54
github.com/k0sproject/k0s/pkg/kubernetes/watch.(*Watcher[...]).Until(...)
        github.com/k0sproject/k0s/pkg/kubernetes/watch/watcher.go:195
github.com/k0sproject/k0s/pkg/component/controller.(*EtcdMemberReconciler).Start.func1()
        github.com/k0sproject/k0s/pkg/component/controller/etcd_member_reconciler.go:109 +0x4a5
created by github.com/k0sproject/k0s/pkg/component/controller.(*EtcdMemberReconciler).Start in goroutine 1
        github.com/k0sproject/k0s/pkg/component/controller/etcd_member_reconciler.go:80 +0x171
```

This occurred after deleting a removed EtcdMember resource via kubectl.